File: Fase3/RESTStudentsPython/src/Student.py
class Student:
        
    def __init__(self, json):           
        self.__firstname = json["firstname"]
        self.__lastname = json["lastname"]
        self.__birthdate = json["birthdate"]
        self.__grades = json["grades"]
        
    def serialize(self):
        return {
            "firstname": self.__firstname,
            "lastname": self.__lastname,
            "birthdate": self.__birthdate,
            "grades": self.__grades,
        }
    
    @property
    def firstname(self):
        return self.__firstname
    
    @firstname.setter
    def firstname(self, firstname):
        self.__firstname = firstname
        
    @property
    def lastname(self):
        return self.__lastname
    
    @lastname.setter
    def lastname(self, lastname):
        self.__lastname = lastname
        
    @property
    def birthdate(self):
        return self.__birthdate
    
    @birthdate.setter
    def birthdate(self, birthdate):
        self.__birthdate = birthdate
        
    @property
    def grades(self):
        return self.__grades
    
    @grades.setter
    def grades(self, grades):
        self.__grades = grades       
        
    def __eq__(self, other):
        return (self.__firstname == other.firstname and 
                self.__lastname == other.lastname and 
                self.__birthdate == other.birthdate)
        
    def __hash__(self):
        return hash(self.__firstname + self.__lastname + self.__birthdate)

File: Fase3/RESTStudentsJava/src/main/java/it/capone/leonardo/rest/students/Student.java
package it.capone.leonardo.rest.students;

import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Objects;

public class Student {

    private String firstname;
    private String lastname;
    private String birthdate;
    private String grades;

    public Student(String firstname, String lastname, String birthdate, String grades) {
        this.firstname = firstname;
        this.lastname = lastname;
        this.birthdate = birthdate;
        this.grades = grades;
    }

    public int age() {
        try {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
            return Period.between(LocalDate.parse(this.birthdate, formatter), LocalDate.now()).getYears();
        }
        catch(DateTimeParseException error) {
            return -1;
        }
    }

    public float avg_grade() {
        String[] grades = this.grades.split(" ");
        float sum = 0;

        try {
            for (String grade : grades) {
                sum += Float.parseFloat(grade);
            }
            return sum / grades.length;
        }
        catch(NumberFormatException error) {
            return -1;
        }
    }

    public String getFirstname() {
        return firstname;
    }

    public void setFirstname(String firstname) {
        this.firstname = firstname;
    }

    public String getLastname() {
        return lastname;
    }

    public void setLastname(String lastname) {
        this.lastname = lastname;
    }

    public String getBirthdate() {
        return birthdate;
    }

    public void setBirthdate(String birthdate) {
        this.birthdate = birthdate;
    }

    public String getGrades() {
        return grades;
    }

    public void setGrades(String grades) {
        this.grades = grades;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Student student = (Student) o;
        return firstname.equals(student.firstname) &&
                lastname.equals(student.lastname) &&
                birthdate.equals(student.birthdate);
    }

    @Override
    public int hashCode() {
        return Objects.hash(firstname, lastname, birthdate);
    }
}
